### PR TITLE
9.0 fix comercial fields

### PR DESCRIPTION
**Pull request by blaggacao**
/cc @kurkop

### CODE REVIEW

**Comment by blaggacao** [comment]
This is originally backported from 10 and it will probably not be put to green.

This changes will be forwarded to 10.0 branch shortly, although there is some new developments on version 10 about the adress topic. We conduct ongoing investigation to gain full compatibility with local requirements, again.

File: l10n_co_geo/models/res_company.py [hunk 1]
```diff
@@ -8,44 +8,55 @@
...
+        if self.country_id:
+            if self.city_id.country_id != self.country_id:
+                self.city_id = False
+            if self.state_id.country_id != self.country_id:
+                self.state_id = False
+        return self._enforce_address_domains()
+
+    @api.model
+    def _address_fields(self):
+        return super(ResCompany, self)._address_fields() + ['city_id']
```
Comment (kurkop) [inline l10n_co_geo/models/res_company.py]: @blaggacao Acá no debería ser + ['city']

File: l10n_co_geo/models/res_company.py [hunk 2]
```diff
@@ -8,44 +8,55 @@
...
+
+    @api.model
```
Comment (kurkop) [inline l10n_co_geo/models/res_company.py]: No debería ser @api.multi?

File: l10n_co_geo/models/res_company.py (hunk 1 above)
Comment (kurkop) [inline l10n_co_geo/models/res_company.py]: @blaggacao Acá no debería ser + ['city']

File: l10n_co_geo/models/res_partner.py [hunk 3]
```diff
@@ -8,22 +8,69 @@
...
+        if self.country_id:
+            domain['state_id'] = [('country_id', '=', self.country_id.id)]
+            domain['city_id'] = [('country_id', '=', self.country_id.id)]
+        if self.state_id:
+            domain['city_id'] = [('state_id', '=', self.state_id.id)]
+        return {'domain': domain}
 
     @api.onchange('city_id')
-    def onchange_city(self):
+    def _onchange_city(self):
```
Comment (kurkop) [inline l10n_co_geo/models/res_partner.py]: No vale la pena hacer un abstract model para incluir estos onchanges en un sólo modelo?.

File: l10n_co_geo/views/res_company_view.xml [hunk 4]
```diff
@@ -1,31 +1,26 @@
...
-            options='{"no_open": True}' on_change="on_change_country(country_id)"
-            invisible="1" />
-        </div>
-      </xpath>
+      <field name="city" position="attributes">
+        <attribute name="invisible">1</attribute>
+      </field>
+      <field name="city" position="after">
+        <field name="city_id" placeholder="City" class="o_address_city" options="{'no_open': True}"
+          domain="['|', ('state_id', '=?', False),'&amp;', ('state_id', '=?', state_id), ('country_id', '=?', country_id)]"
```
Comment (kurkop) [inline l10n_co_geo/views/res_company_view.xml]: Qué significa '=?', estos dominios no entran en conflicto con los del onchange?

File: l10n_co_geo/views/res_partner_view.xml [hunk 5]
```diff
@@ -4,64 +4,49 @@
...
       </xpath>
-      <xpath expr="//field[@name='child_ids']" position="attributes">
+      <field name="state_id" position="attributes">
+        <attribute name="on_change" />
+        <attribute name="domain">[('country_id', '=', country_id or True)]</attribute>
+        <attribute name="context">{'default_country_id': country_id}</attribute>
+      </field>
+      <xpath expr="//notebook/page[1]/field/form/sheet//field[@name='state_id']" position="attributes">
+        <attribute name="on_change" />
+        <attribute name="domain">[('country_id', '=', country_id or True)]</attribute>
```
Comment (kurkop) [inline l10n_co_geo/views/res_partner_view.xml]: No entiendo el dominio country_id or True, esto no sería siempre True?

Comment (blaggacao) [inline l10n_co_geo/views/res_partner_view.xml]:
```python
>>> country_id = 1
>>> country_id_model = 5
>>> country_id == True
True
>>> country_id == country_id_model
False
```

File: l10n_co_geo/views/res_company_view.xml (hunk 4 above)
Comment (blaggacao) [inline l10n_co_geo/views/res_company_view.xml]:
```python
('state_id', '=?', state_id) == ('state_id', '=', state_id or False)
```

File: l10n_co_geo/models/res_partner.py (hunk 3 above)
Comment (blaggacao) [inline l10n_co_geo/models/res_partner.py]: dicho. hecho. 💃

File: l10n_co_geo/models/res_company.py (hunk 2 above)
Comment (blaggacao) [inline l10n_co_geo/models/res_company.py]: `create` is a `model` method, similar to the `@classmethod` decorator.
`@api.model` = kind of an "odoo-classmethod"

Comment (blaggacao) [inline l10n_co_geo/models/res_company.py]: writes, constraints, and others are multies indeed. Little known fact: onchanges are always executed on a single ! pseudo record. (no need for `self.ensure_one()` on them)

File: l10n_co_geo/models/res_company.py (hunk 1 above)
Comment (blaggacao) [inline l10n_co_geo/models/res_company.py]: nope, that's the whole point about it: `city_id`

File: l10n_co_geo/models/res_partner.py (hunk 3 above)
Comment (blaggacao) [inline l10n_co_geo/models/res_partner.py]: sidenote `super` calls are not abstract model compatible... So they where left on their corresponding model, everything else is abstract now, and I even included an improvement through consistency on the actual business logic..

File: l10n_co_geo/views/res_partner_view.xml (hunk 5 above)
Comment (blaggacao) [inline l10n_co_geo/views/res_partner_view.xml]: Tienes razon, esto no tiene sentido, creo que queria decir `country_id or True` o como arriba `=?`

Comment (blaggacao) [inline l10n_co_geo/views/res_partner_view.xml]: Ya recupere: no, no tienes razon. `(country_id or True)` evalua a dos resultados distintos que cada uno por si mismo forma parte del leaf en el respectivo caso. Lo que escribí corresponde exactamente al operador `'=?'`... No me confundes! 😄

Comment (blaggacao) [inline l10n_co_geo/views/res_partner_view.xml]: No siendo incorrecto, era mal estilo, lo cambie al mas idiomatico `[('country_id', '=?', country_id)]`

File: l10n_co_geo/views/res_company_view.xml (hunk 4 above)
Comment (blaggacao) [inline l10n_co_geo/views/res_company_view.xml]: Efectivamente era un legado de la forma como estaba anteriormente, corregido:
`"[('state_id', '=?', state_id), ('country_id', '=?', country_id)]"`, no mas.

**Pull request closed without merging**